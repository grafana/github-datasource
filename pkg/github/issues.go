package github

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/grafana/github-datasource/pkg/models"
	"github.com/grafana/grafana-plugin-sdk-go/data"
	"github.com/pkg/errors"
	"github.com/shurcooL/githubv4"
)

// Issue represents a GitHub issue in a repository
type Issue struct {
	Title     string
	ClosedAt  githubv4.DateTime
	CreatedAt githubv4.DateTime
	Closed    bool
	Author    struct {
		User `graphql:"... on User"`
	}
}

// Issues is a slice of GitHub issues
type Issues []Issue

// Frames converts the list of issues to a Grafana DataFrame
func (c Issues) Frames() data.Frames {
	frame := data.NewFrame(
		"issues",
		data.NewField("title", nil, []string{}),
		data.NewField("author", nil, []string{}),
		data.NewField("author_company", nil, []string{}),
		data.NewField("closed", nil, []bool{}),
		data.NewField("created_at", nil, []time.Time{}),
		data.NewField("closed_at", nil, []*time.Time{}),
	)

	for _, v := range c {
		var closedAt *time.Time
		if !v.ClosedAt.Time.IsZero() {
			closedAt = &v.ClosedAt.Time
		}

		frame.AppendRow(
			v.Title,
			v.Author.User.Login,
			v.Author.User.Company,
			v.Closed,
			v.CreatedAt.Time,
			closedAt,
		)
	}

	return data.Frames{frame}
}

// QuerySearchIssues is the object representation of the graphql query for retrieving a paginated list of issues using the search query
// {
//   search(query: "is:issue repo:grafana/grafana opened:2020-08-19..*", type: ISSUE, first: 100) {
//     nodes {
//       ... on PullRequest {
//         id
//         title
//       }
//   }
// }
type QuerySearchIssues struct {
	Search struct {
		Nodes []struct {
			Issue Issue `graphql:"... on Issue"`
		}
		PageInfo PageInfo
	} `graphql:"search(query: $query, type: ISSUE, first: 100, after: $cursor)"`
}

// GetIssuesInRange lists issues in a project given a time range.
func GetIssuesInRange(ctx context.Context, client Client, opts models.ListIssuesOptions, from time.Time, to time.Time) (Issues, error) {
	search := []string{
		"is:issue",
		fmt.Sprintf("repo:%s/%s", opts.Owner, opts.Repository),
		fmt.Sprintf("%s:%s..%s", opts.TimeField.String(), from.Format(time.RFC3339), to.Format(time.RFC3339)),
	}

	if opts.Query != nil {
		search = append(search, *opts.Query)
	}

	var (
		variables = map[string]interface{}{
			"cursor": (*githubv4.String)(nil),
			"query":  githubv4.String(strings.Join(search, " ")),
		}

		issues = []Issue{}
	)

	for {
		q := &QuerySearchIssues{}
		if err := client.Query(ctx, q, variables); err != nil {
			return nil, errors.WithStack(err)
		}
		is := make([]Issue, len(q.Search.Nodes))

		for i, v := range q.Search.Nodes {
			is[i] = v.Issue
		}

		issues = append(issues, is...)

		if !q.Search.PageInfo.HasNextPage {
			break
		}
		variables["cursor"] = q.Search.PageInfo.EndCursor
	}

	return issues, nil
}
