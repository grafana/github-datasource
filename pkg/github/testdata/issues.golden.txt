ðŸŒŸ This was machine generated.  Do not edit. ðŸŒŸ

Frame[0] 
Name: issues
Dimensions: 8 Fields by 3 Rows
+----------------+----------------+----------------------+-----------------+---------------+--------------+-------------------------------+-------------------------------+
| Name: title    | Name: author   | Name: author_company | Name: repo      | Name: number  | Name: closed | Name: created_at              | Name: closed_at               |
| Labels:        | Labels:        | Labels:              | Labels:         | Labels:       | Labels:      | Labels:                       | Labels:                       |
| Type: []string | Type: []string | Type: []string       | Type: []string  | Type: []int64 | Type: []bool | Type: []time.Time             | Type: []*time.Time            |
+----------------+----------------+----------------------+-----------------+---------------+--------------+-------------------------------+-------------------------------+
| Issue #1       | firstUser      | ACME Corp            | grafana/grafana | 1             | false        | 2020-08-25 16:21:56 +0000 UTC | null                          |
| Issue #2       | secondUser     | ACME Corp            | grafana/grafana | 2             | true         | 2020-08-25 16:21:56 +0000 UTC | 2020-08-25 22:21:56 +0000 UTC |
| Issue #3       | firstUser      | ACME Corp            | grafana/grafana | 3             | false        | 2020-08-25 16:21:56 +0000 UTC | null                          |
+----------------+----------------+----------------------+-----------------+---------------+--------------+-------------------------------+-------------------------------+


====== TEST DATA RESPONSE (arrow base64) ======
FRAME=QVJST1cxAAD/////4AMAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAFQAAAACAAAAKAAAAAQAAACo/P//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAAMj8//8IAAAAEAAAAAYAAABpc3N1ZXMAAAQAAABuYW1lAAAAAAgAAADwAgAAgAIAABQCAAC4AQAATAEAAPAAAACAAAAAGAAAAAAAEgAYABQAEwASAAwAAAAIAAQAEgAAABQAAABAAAAAQAAAAAAACgFAAAAAAQAAAAQAAABE/f//CAAAABQAAAAJAAAAY2xvc2VkX2F0AAAABAAAAG5hbWUAAAAAAAAAAJr///8AAAMACQAAAGNsb3NlZF9hdAAAALr9//8UAAAAQAAAAEgAAAAAAAAKSAAAAAEAAAAEAAAAqP3//wgAAAAUAAAACgAAAGNyZWF0ZWRfYXQAAAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMACgAAAGNyZWF0ZWRfYXQAACb+//8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAAFP7//wgAAAAQAAAABgAAAGNsb3NlZAAABAAAAG5hbWUAAAAAAAAAAAz+//8GAAAAY2xvc2VkAAB+/v//FAAAADwAAABEAAAAAAAAAkgAAAABAAAABAAAAGz+//8IAAAAEAAAAAYAAABudW1iZXIAAAQAAABuYW1lAAAAAAAAAAAIAAwACAAHAAgAAAAAAAABQAAAAAYAAABudW1iZXIAAOb+//8UAAAAPAAAADwAAAAAAAAFOAAAAAEAAAAEAAAA1P7//wgAAAAQAAAABAAAAHJlcG8AAAAABAAAAG5hbWUAAAAAAAAAAMz+//8EAAAAcmVwbwAAAAA+////FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAACz///8IAAAAGAAAAA4AAABhdXRob3JfY29tcGFueQAABAAAAG5hbWUAAAAAAAAAACz///8OAAAAYXV0aG9yX2NvbXBhbnkAAKb///8UAAAAPAAAADwAAAAAAAAFOAAAAAEAAAAEAAAAlP///wgAAAAQAAAABgAAAGF1dGhvcgAABAAAAG5hbWUAAAAAAAAAAIz///8GAAAAYXV0aG9yAAAAABIAGAAUAAAAEwAMAAAACAAEABIAAAAUAAAARAAAAEgAAAAAAAAFRAAAAAEAAAAMAAAACAAMAAgABAAIAAAACAAAABAAAAAFAAAAdGl0bGUAAAAEAAAAbmFtZQAAAAAAAAAABAAEAAQAAAAFAAAAdGl0bGUAAAAAAAAA/////xgCAAAUAAAAAAAAAAwAFgAUABMADAAEAAwAAAAgAQAAAAAAABQAAAAAAAADAwAKABgADAAIAAQACgAAABQAAABYAQAAAwAAAAAAAAAAAAAAFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAYAAAAAAAAACgAAAAAAAAAAAAAAAAAAAAoAAAAAAAAABAAAAAAAAAAOAAAAAAAAAAgAAAAAAAAAFgAAAAAAAAAAAAAAAAAAABYAAAAAAAAABAAAAAAAAAAaAAAAAAAAAAgAAAAAAAAAIgAAAAAAAAAAAAAAAAAAACIAAAAAAAAABAAAAAAAAAAmAAAAAAAAAAwAAAAAAAAAMgAAAAAAAAAAAAAAAAAAADIAAAAAAAAABgAAAAAAAAA4AAAAAAAAAAAAAAAAAAAAOAAAAAAAAAACAAAAAAAAADoAAAAAAAAAAAAAAAAAAAA6AAAAAAAAAAYAAAAAAAAAAABAAAAAAAACAAAAAAAAAAIAQAAAAAAABgAAAAAAAAAAAAAAAgAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAACAAAAAAAAAAAAAAAIAAAAEAAAABgAAABJc3N1ZSAjMUlzc3VlICMySXNzdWUgIzMAAAAACQAAABMAAAAcAAAAZmlyc3RVc2Vyc2Vjb25kVXNlcmZpcnN0VXNlcgAAAAAAAAAACQAAABIAAAAbAAAAQUNNRSBDb3JwQUNNRSBDb3JwQUNNRSBDb3JwAAAAAAAAAAAADwAAAB4AAAAtAAAAZ3JhZmFuYS9ncmFmYW5hZ3JhZmFuYS9ncmFmYW5hZ3JhZmFuYS9ncmFmYW5hAAAAAQAAAAAAAAACAAAAAAAAAAMAAAAAAAAAAgAAAAAAAAAAaO2yVY8uFgBo7bJVjy4WAGjtslWPLhYCAAAAAAAAAAAAAAAAAAAAAChB1/qiLhYAAAAAAAAAABAAAAAMABQAEgAMAAgABAAMAAAAEAAAACwAAAA4AAAAAAADAAEAAADwAwAAAAAAACACAAAAAAAAIAEAAAAAAAAAAAAAAAAAAAAACgAMAAAACAAEAAoAAAAIAAAAVAAAAAIAAAAoAAAABAAAAKj8//8IAAAADAAAAAAAAAAAAAAABQAAAHJlZklkAAAAyPz//wgAAAAQAAAABgAAAGlzc3VlcwAABAAAAG5hbWUAAAAACAAAAPACAACAAgAAFAIAALgBAABMAQAA8AAAAIAAAAAYAAAAAAASABgAFAATABIADAAAAAgABAASAAAAFAAAAEAAAABAAAAAAAAKAUAAAAABAAAABAAAAET9//8IAAAAFAAAAAkAAABjbG9zZWRfYXQAAAAEAAAAbmFtZQAAAAAAAAAAmv///wAAAwAJAAAAY2xvc2VkX2F0AAAAuv3//xQAAABAAAAASAAAAAAAAApIAAAAAQAAAAQAAACo/f//CAAAABQAAAAKAAAAY3JlYXRlZF9hdAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAKAAAAY3JlYXRlZF9hdAAAJv7//xQAAAA8AAAAPAAAAAAAAAY4AAAAAQAAAAQAAAAU/v//CAAAABAAAAAGAAAAY2xvc2VkAAAEAAAAbmFtZQAAAAAAAAAADP7//wYAAABjbG9zZWQAAH7+//8UAAAAPAAAAEQAAAAAAAACSAAAAAEAAAAEAAAAbP7//wgAAAAQAAAABgAAAG51bWJlcgAABAAAAG5hbWUAAAAAAAAAAAgADAAIAAcACAAAAAAAAAFAAAAABgAAAG51bWJlcgAA5v7//xQAAAA8AAAAPAAAAAAAAAU4AAAAAQAAAAQAAADU/v//CAAAABAAAAAEAAAAcmVwbwAAAAAEAAAAbmFtZQAAAAAAAAAAzP7//wQAAAByZXBvAAAAAD7///8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAALP///wgAAAAYAAAADgAAAGF1dGhvcl9jb21wYW55AAAEAAAAbmFtZQAAAAAAAAAALP///w4AAABhdXRob3JfY29tcGFueQAApv///xQAAAA8AAAAPAAAAAAAAAU4AAAAAQAAAAQAAACU////CAAAABAAAAAGAAAAYXV0aG9yAAAEAAAAbmFtZQAAAAAAAAAAjP///wYAAABhdXRob3IAAAAAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAABEAAAASAAAAAAAAAVEAAAAAQAAAAwAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAUAAAB0aXRsZQAAAAQAAABuYW1lAAAAAAAAAAAEAAQABAAAAAUAAAB0aXRsZQAAAAgEAABBUlJPVzE=
